pkg/evaluation/evaluator.go

```diff
@@ -0,0 +1,355 @@
+	equalSensitiveOperator = "equal_sensitive"
+)
+
+// Query provides methods for segment and flag retrieval
+type Query interface {
+	GetSegment(identifier string) (rest.Segment, error)
+	GetFlag(identifier string) (rest.FeatureConfig, error)
+}
+
+// Evaluator engine evaluates flag from provided query
+type Evaluator struct {
+	query Query
+}
+
+// NewEvaluator constructs evaluator with query instance
+func NewEvaluator(query Query) *Evaluator {
+	return &Evaluator{
+		query: query,
+	}
+}
+
+func (e Evaluator) evaluateClause(clause *rest.Clause, target *rest.Target) bool {
+	if clause == nil {
+		return false
+	}
+
+	operator := clause.Op
+	if operator == "" {
+		return false
+	}
+
+	attrValue := getAttrValue(target, clause.Attribute)
+	if attrValue.IsNil() {
+		return false
+	}
+
+	object := attrValue.String()
+	values := clause.Values
+	if len(values) == 0 {
+		return false
+	}
+
+	value := values[0]
+
+	switch operator {
+	case startsWithOperator:
+		return strings.HasPrefix(object, value)
+	case endsWithOperator:
+		return strings.HasSuffix(object, value)
+	case matchOperator:
+		found, err := regexp.MatchString(value, object)
+		if err != nil || !found {
+			return false
+		}
+		return true
+	case containsOperator:
+		return strings.Contains(object, value)
+	case equalOperator:
+		return strings.EqualFold(object, value)
+	case equalSensitiveOperator:
+		return object == value
+	case inOperator:
+		for _, val := range values {
+			if val == object {
+				return true
+			}
+		}
+		return false
+	case gtOperator:
+		return object > value
+	case segmentMatchOperator:
+		return e.isTargetIncludedOrExcludedInSegment(&values, target)
+	default:
+		return false
+	}
+}
+
+func (e Evaluator) evaluateClauses(clauses []rest.Clause, target *rest.Target) bool {
+	for i := range clauses {
+		if !e.evaluateClause(&clauses[i], target) {
+			return false
+		}
+	}
+	return true
+}
+
+func (e Evaluator) evaluateRule(servingRule *rest.ServingRule, target *rest.Target) bool {
+	return e.evaluateClauses(servingRule.Clauses, target)
+}
+
+func (e Evaluator) evaluateRules(servingRules *[]rest.ServingRule, target *rest.Target) string {
+	if target == nil || servingRules == nil {
+		return ""
+	}
+
+	sort.SliceStable(*servingRules, func(i, j int) bool {
+		return (*servingRules)[i].Priority < (*servingRules)[j].Priority
+	})
+	for i := range *servingRules {
+		rule := (*servingRules)[i]
+		// if evaluation is false just continue to next rule
+		if !e.evaluateRule(&rule, target) {
+			continue
+		}
+
+		// rule matched, check if there is distribution
+		if rule.Serve.Distribution != nil {
+			return evaluateDistribution(rule.Serve.Distribution, target)
+		}
+
+		// rule matched, here must be variation if distribution is undefined or null
+		if rule.Serve.Variation != nil {
+			return *rule.Serve.Variation
+		}
+	}
+	return ""
+}
+
+func (e Evaluator) evaluateVariationMap(variationsMap *[]rest.VariationMap, target *rest.Target) string {
```

Comment:
I'm not sure this makes sense, as a slice is just a header, so you could avoid the derefernece a nil check here.

Reply:
I just followed rest types, Yes I know about slicing and underlaying arrays